prow/metrics/metrics_test.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/test-infra/prow/config"
+)
+
+func TestMetricsPusherIntervals(t *testing.T) {
+	intervals := []time.Duration{15 * time.Millisecond, 0 * time.Millisecond, 10 * time.Millisecond}
+	lastRunTime := time.Now()
+	ranFuncCh := make(chan time.Time)
+
+	intervalCheckGatherer := func(job string, grouping map[string]string, url string, g prometheus.Gatherer) error {
+		ranFuncCh <- time.Now()
+		return nil
+	}
+
+	pusher := Pusher{
+		gatherer: intervalCheckGatherer,
+		configAgent: &fakeConfigAgent{
+			intervals: intervals,
+		},
+	}
+	go pusher.Start("")
+
+	for _, expectedInterval := range intervals {
```

Comment:
I'm afraid a timing-based test with allowable error in the range of milliseconds will flake relentlessly in CI if it's run in parallel with other workloads or gets preempted or whatever else happens.